pandas/core/internals/managers.py

```diff
@@ -803,19 +809,34 @@ def as_array(self, transpose: bool = False) -> np.ndarray:
             # always be object dtype. Some callers seem to want the
             # DatetimeArray (previously DTI)
             arr = self.blocks[0].get_values(dtype=object)
+        elif self._is_single_block and self.blocks[0].is_extension:
+            # Avoid implicit conversion of extension blocks to object
+            arr = (
+                self.blocks[0]
+                .values.to_numpy(dtype=dtype, na_value=na_value)
+                .reshape(self.blocks[0].shape)
+            )
         elif self._is_single_block or not self.is_mixed_type:
             arr = np.asarray(self.blocks[0].get_values())
+            if dtype:
+                arr = arr.astype(dtype)
         else:
-            arr = self._interleave()
+            arr = self._interleave(dtype=dtype, na_value=na_value)
+
+        if na_value is not lib.no_default:
+            na_mask = isna(arr)
+            if na_mask.any():
+                arr[na_mask] = na_value
```

Comment:
Actually, if ``arr`` is not a copy at this point, this will modify the original dataframe (eg I assume this will happen with a uniform float dataframe and you do eg `df.to_numpy(na_value=0)`). Can you check if that indeed is the case and add a test for this to ensure the original dataframe didn't change in the single block case?

Reply:
That's a good point, it looks like the original DataFrame is _not_ mutated (at least in the case you mention)

Reply:
Now that we may be copying multiple times, I think we should probably move the copy into this function and make sure that doesn't happen?

Reply:
Yes, that's probably a good idea